pretyping/pretyping.ml

```diff
@@ -927,7 +927,61 @@ struct
             sigma, na.binder_name, Some dom, Some rng
           else
             sigma, Anonymous, None, None
+        | Proj (proj, arg) when isEvar sigma arg ->
+          begin
+            try
+              (* Quick check to see if this can succeed at all, avoiding a costly unification call if possible. *)
+              let _ =
+                try CanonicalSolution.find (GlobEnv.env env) sigma (Names.GlobRef.ConstRef (Projection.constant proj), ValuePattern.Prod_cs)
+                with Not_found ->
+                  CanonicalSolution.find (GlobEnv.env env) sigma (Names.GlobRef.ConstRef (Projection.constant proj), ValuePattern.Default_cs)
+              in
+              let sigma, (domain_hole, _) = Evarutil.new_type_evar (GlobEnv.env env) sigma Evd.univ_flexible in
+              let sigma, (codomain_hole, _) = Evarutil.new_type_evar (GlobEnv.env env) sigma Evd.univ_flexible in
+              let binder = Context.anonR in
+              let cs_type = mkProd (binder, domain_hole, codomain_hole) in
+
+              let sigma = Evarconv.unify
+                  (GlobEnv.env env)
+                  sigma
+                  Reduction.CUMUL
+                  ty
+                  cs_type
+              in
+              sigma, Context.binder_name binder, Some domain_hole, Some codomain_hole
+            with
+            | Not_found | PretypeError _ -> error_not_product ?loc (GlobEnv.env env) sigma ty
+            end
         | _ ->
+          try
+            let (proj, inst, evar, params) = CanonicalSolution.decompose_reduced_canonical_projection sigma ty in
```

Comment:
Why is `evar` called that? Is there a missing `isEvar`?

Reply:
I think I assumed that a) `ty` was in hnf and that b) the only possible case for an hnf term to be a match (necessary for `decompose_reduced_canonical_projection` to succeed) is that the discriminee is an evar. I see now that neither assumption is true. (I don't think the type is in hnf in coercion.ml and the match could be stuck because of an opaque term.) I see that you already changed this so I assume this comment is no longer relevant?

Reply:
> I see that you already changed this so I assume this comment is no longer relevant?

That depends on if the assumption was used only to decide variable naming or if it has deeper implications.

Reply:
I am trying to investigate this but I can't really make heads or tails of it. I think we probably want an `isEvar` check here but I can't say that I know of a situation in which it would fail.